scripts/game_structure/game_essentials.py

```diff
@@ -191,6 +193,50 @@ def update_game(self):
             self.switch_screens = True
         self.clicked = False
         self.keyspressed = []
+    
+    @staticmethod
+    def safe_save(path: str, write_data, max_attempts: int=15):
+        """ Attempt to safely safe a file.
+            If write_data is not a string, assumes you want this
+            in json format. 
+            Raises a RunTimeError is 
+            the file is nullied too many times. """
+        
+        # If write_data is not a string, 
+        if type(write_data) is not str:
+            _data = ujson.dumps(write_data, indent=4)
+        else:
+            _data = write_data
+            
+        dir_name, file_name = os.path.split(path)
+        
+        temp_file_path = get_temp_dir() + "/" + file_name + ".tmp"
+        i = 0
+        while True:
+            # Attempt to write to temp file
+            with open(temp_file_path, "w") as write_file:
+                write_file.write(_data)
+            # Read the file back in, in binary mode, so
+            # we can properly check for file nullification. 
+            # We don't need to read the entire file, 
+            # so let's just read the first 200 bytes. 
+            with open(temp_file_path, 'rb') as read_file:
+                _binary_data = read_file.read(200)
+            
+            # If_data is not empty, and _binary_data is
+            # all null_bytes or empty. 
+            if _data and not any(_binary_data):
+                i += 1
+                if i > max_attempts:
+                    print(f"{file_name} has been nullied too many times. Saving Failed.")
+                    raise RuntimeError(f"{file_name} was nullied too many times!")
+                print(f"{file_name} was nullifed. Trying again.")
+                continue
```

Comment:
To verify the integrity of the file, can't we just read it back entirely and check if `write_data` equals whatever was written in the file? That might be a bit more inefficient, but _guarantees_ that the file is saved correctly.

Reply:
That's a good point. I originally only read part of the file for performance reasons, but some quick tests with a large save show the decrease in performance is minimal. I have changed the function to read in the entire file.